### PR TITLE
Get PolynomialDecomposer to Parity with STLDecomposer

### PR DESCRIPTION
**Pull request by chukarsten**
A follow up PR to move more of the tests over to `test_decomposer.py` and parametrize them between the two child decomposers in order to prepare for more decomposers.

### CODE REVIEW

File: evalml/tests/component_tests/decomposer_tests/test_decomposer.py [hunk 1]
```diff
@@ -16,6 +16,10 @@
 decomposer_list = [STLDecomposer, PolynomialDecomposer]
 
 
+def get_trend_dataframe_format_correct(df):
```
Comment (chukarsten) [inline evalml/tests/component_tests/decomposer_tests/test_decomposer.py:19]: Just moved this up from the child class test modules.

File: evalml/tests/component_tests/decomposer_tests/test_decomposer.py [hunk 2]
```diff
@@ -34,6 +38,21 @@ def test_set_time_index(decomposer_child_class):
     assert isinstance(y_time_index.index, pd.DatetimeIndex)
 
 
+@pytest.mark.parametrize(
```
Comment (chukarsten) [inline evalml/tests/component_tests/decomposer_tests/test_decomposer.py:41]: Moved all this up from child class test modules.

File: evalml/tests/component_tests/decomposer_tests/test_decomposer.py [hunk 3]
```diff
@@ -481,3 +500,125 @@ def test_decomposer_bad_target_index(
         match="doesn't support target data with index of type",
     ):
         dec._choose_proper_index(y)
+
+
+@pytest.mark.parametrize(
```
Comment (chukarsten) [inline evalml/tests/component_tests/decomposer_tests/test_decomposer.py:505]: Moved all this up from child class test modules....

File: evalml/tests/component_tests/decomposer_tests/test_decomposer.py [hunk 4]
```diff
@@ -481,3 +500,125 @@ def test_decomposer_bad_target_index(
...
+            ):
+                output_X, output_inverse_y = decomposer.transform(None, y_new)
+        else:
+            output_X, output_y_t = decomposer.transform(None, y[y_new.index])
+
+            pd.testing.assert_series_equal(
+                pd.Series(np.zeros(len(output_y_t))).set_axis(y_new.index),
+                output_y_t,
+                check_exact=False,
+                atol=0.1,
```
Comment (chukarsten) [inline evalml/tests/component_tests/decomposer_tests/test_decomposer.py:573]: I did have to loosen this tolerance (`atol`) here a bit.  Part of the problem was that the different decomposers are able to more or less properly capture the trend.  Based on how much of the trend they capture, the answers will be closer or further away from a horizontal line.  Maybe this isn't a good way to test the transform...you tell me!

File: evalml/tests/component_tests/decomposer_tests/test_decomposer.py [hunk 5]
```diff
@@ -481,3 +500,125 @@ def test_decomposer_bad_target_index(
...
+    decomposer.fit(subset_X, subset_y)
+
+    if transformer_fit_on_data == "in-sample":
+        output_X, output_y = decomposer.transform(subset_X, subset_y)
+        pd.testing.assert_series_equal(
+            pd.Series(np.zeros(len(output_y))).set_axis(subset_y.index),
+            output_y,
+            check_dtype=False,
+            check_names=False,
+            atol=0.2,
```
Comment (chukarsten) [inline evalml/tests/component_tests/decomposer_tests/test_decomposer.py:547]: This tolerance was still good enough for both decomposers.

File: evalml/tests/component_tests/decomposer_tests/test_decomposer.py [hunk 6]
```diff
@@ -481,3 +500,125 @@ def test_decomposer_bad_target_index(
...
+    generate_seasonal_data,
+    transformer_fit_on_data,
+):
+    # Generate 10 periods (the default) of synthetic seasonal data
+    seasonal_period = 7
+    X, y = generate_seasonal_data(real_or_synthetic="synthetic")(
+        period=seasonal_period,
+        freq_str="D",
+        set_time_index=True,
+        seasonal_scale=0.05,  # Increasing this value causes the decomposer to miscalculate trend
```
Comment (chukarsten) [inline evalml/tests/component_tests/decomposer_tests/test_decomposer.py:532]: I added this to try and get the decomposer to perform a little better and do a better job capturing the trend.  This just reduces the amplitude of the sinusoidal oscillations added to the synthetic signal.

File: evalml/tests/component_tests/decomposer_tests/test_decomposer.py [hunk 7]
```diff
@@ -481,3 +500,125 @@ def test_decomposer_bad_target_index(
...
+
+            pd.testing.assert_series_equal(
+                pd.Series(np.zeros(len(output_y_t))).set_axis(y_new.index),
+                output_y_t,
+                check_exact=False,
+                atol=0.1,
+            )
+
+
+def build_test_target(subset_y, seasonal_period, transformer_fit_on_data, to_test):
```
Comment (chukarsten) [inline evalml/tests/component_tests/decomposer_tests/test_decomposer.py:577]: This function has been moved up to the base test module class unchanged.

File: evalml/pipelines/components/transformers/preprocessing/polynomial_decomposer.py [hunk 8]
```diff
@@ -180,20 +185,80 @@ def inverse_transform(self, y_t: pd.Series) -> tuple[pd.DataFrame, pd.Series]:
...
+            y_t_in_sample = y_t[y_t_ind]
+
+            # Convert y_t to datetime index to use the built in component
+            if isinstance(y_t.index, (pd.RangeIndex, Int64Index)):
+                y_t_dt_ind = self._convert_int_index_to_dt_index(y_t_in_sample.index)
+            elif isinstance(y_t.index, pd.DatetimeIndex):
+                y_t_dt_ind = y_t_ind
+            trend = self._component_obj.inverse_transform(
+                y_t_in_sample.set_axis(y_t_dt_ind),
+            )
```
Comment (chukarsten) [inline evalml/pipelines/components/transformers/preprocessing/polynomial_decomposer.py:225]: This has been the main offender and is one of the main differences between the `STLDecomposer` and `PolynomialDecomposer`.  With respect to the trend, the PD's underlying object has the same fit/transform/inverse_transform API that our components do, so it fits well.  The STLD does not have this so requires a lot of code to do manual trend projection.  I believe that both required the detrended target signal to have datetime indices, which is challenging.

File: evalml/tests/component_tests/decomposer_tests/test_decomposer.py (hunk 1 above)
Comment (ParthivNaresh) [reply evalml/tests/component_tests/decomposer_tests/test_decomposer.py:19]: What is this test testing?

File: evalml/tests/component_tests/decomposer_tests/test_decomposer.py (hunk 6 above)
Comment (ParthivNaresh) [reply evalml/tests/component_tests/decomposer_tests/test_decomposer.py:537]: Why does increasing seasonal_scale cause the decomposer to miscalculate the trend?

File: evalml/pipelines/components/transformers/preprocessing/polynomial_decomposer.py [hunk 9]
```diff
@@ -95,6 +95,7 @@ def fit(self, X: pd.DataFrame, y: pd.Series = None) -> PolynomialDecomposer:
         """
         self.original_index = y.index if y is not None else None
         X, y = self._check_target(X, y)
+        self._map_dt_to_integer(self.original_index, y.index)
```
[the remark this answers is not in the document]
Comment (ParthivNaresh) [reply evalml/pipelines/components/transformers/preprocessing/polynomial_decomposer.py:98]: `_map_dt_to_integer` is called in both STL and PD but `_convert_int_index_to_dt_index` is only called for the PD? What's the reason for this?

**Review by ParthivNaresh — APPROVED** [review]
Left some random questions but I think overall it seems to be on track, nice work!

File: evalml/pipelines/components/transformers/preprocessing/polynomial_decomposer.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (chukarsten) [reply evalml/pipelines/components/transformers/preprocessing/polynomial_decomposer.py:98]: Yea, to be fair, it's probably not needed in STL.  Yet.  It kind of depends on how we want to move forward with standardizing the fit functions and what attributes we want to be defined in the class.  I sort of want to do an additional PR to completely standardize both fit functions and see if I can pass that up into.

File: evalml/tests/component_tests/decomposer_tests/test_decomposer.py (hunk 1 above)
Comment (chukarsten) [reply evalml/tests/component_tests/decomposer_tests/test_decomposer.py:19]: It's more a utility for the tests to use, just to make sure that `get_trend_df()` returns a dataframe with the proper columns.  `get_trend_dataframe_format_correct()` will be used in the `get_trend_dataframe()` tests.

File: evalml/tests/component_tests/decomposer_tests/test_decomposer.py (hunk 6 above)
Comment (chukarsten) [reply evalml/tests/component_tests/decomposer_tests/test_decomposer.py:537]: That has yet to be determined.  What we've noticed in some datasets is that there is a tendency for the periodic signal to "leak" into the trend.  Loosely, I've noticed that a larger relative seasonal signal vs. trend exacerbates this to a degree.

File: evalml/pipelines/components/transformers/preprocessing/polynomial_decomposer.py (hunk 8 above)
Comment (jeremyliweishih) [reply evalml/pipelines/components/transformers/preprocessing/polynomial_decomposer.py:225]: just to clarify - this new block ensures that both STL and PD are using the same logic when calculating `inverse_transform` and not depend on PD's native `inverse_transform`?

**Review by jeremyliweishih — APPROVED** [review]
This LGTM - seeing the tests pass like this gives me good confidence in parity 👍

**Pull request merged**